pkg/api/file.go

```diff
@@ -198,6 +200,143 @@ func (s *server) fileUploadHandler(w http.ResponseWriter, r *http.Request) {
+		}
+
+		// try to find filename
+		// 1) in part header params
+		// 2) as formname
+		// 3) file reference hash (after uploading the file)
+		if fileName = part.FileName(); fileName == "" {
+			fileName = part.FormName()
+		}
+
+		// then find out content type
+		contentType = part.Header.Get("Content-Type")
+		if contentType == "" {
+			br := bufio.NewReader(part)
+			buf, err := br.Peek(512)
+			if err != nil && err != io.EOF {
+				return nil, fmt.Errorf("read content type error: %v", err)
+			}
+			contentType = http.DetectContentType(buf)
+			reader = br
+		} else {
+			reader = part
+		}
+		contentLength = part.Header.Get("Content-Length")
+	} else {
+		fileName = r.URL.Query().Get("name")
+		contentLength = r.Header.Get("Content-Length")
+		reader = r.Body
+	}
+
+	if contentLength != "" {
+		fileSize, err = strconv.ParseUint(contentLength, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("invalid content length error: %v", err)
+		}
+	} else {
+		// copy the part to a tmp file to get its size
+		tmp, err := ioutil.TempFile("", "bee-multipart")
+		if err != nil {
+			return nil, fmt.Errorf("create temp file error: %v", err)
+		}
+		defer os.Remove(tmp.Name())
+		defer tmp.Close()
+		n, err := io.Copy(tmp, reader)
+		if err != nil {
+			return nil, fmt.Errorf("write temp file error: %v", err)
+		}
+		if _, err := tmp.Seek(0, io.SeekStart); err != nil {
+			return nil, fmt.Errorf("seek temp file error: %v", err)
+		}
+		fileSize = uint64(n)
+		reader = tmp
+	}
+
+	return &FileUploadInfo{
+		name:        fileName,
+		size:        int64(fileSize),
+		contentType: contentType,
+		toEncrypt:   toEncrypt,
+		reader:      reader,
+	}, nil
+}
+
+// storeFile uploads the given file and returns its reference
+// this function was extracted from `fileUploadHandler` and should eventually replace its current code, along with `GetFileHTTPInfo`
+func storeFile(ctx context.Context, fileInfo *FileUploadInfo, s storage.Storer) (swarm.Address, error) {
+	// first store the file and get its reference
+	sp := splitter.NewSimpleSplitter(s)
+	fr, err := file.SplitWriteAll(ctx, sp, fileInfo.reader, fileInfo.size, fileInfo.toEncrypt)
+	if err != nil {
+		return swarm.ZeroAddress, fmt.Errorf("split file error: %v", err)
+	}
+
+	// if filename is still empty, use the file hash as the filename
+	if fileInfo.name == "" {
+		fileInfo.name = fr.String()
+	}
+
+	// then store the metadata and get its reference
+	m := entry.NewMetadata(fileInfo.name)
+	m.MimeType = fileInfo.contentType
+	metadataBytes, err := json.Marshal(m)
+	if err != nil {
+		return swarm.ZeroAddress, fmt.Errorf("metadata marshal error: %v", err)
+	}
+
+	sp = splitter.NewSimpleSplitter(s)
+	mr, err := file.SplitWriteAll(ctx, sp, bytes.NewReader(metadataBytes), int64(len(metadataBytes)), fileInfo.toEncrypt)
+	if err != nil {
+		return swarm.ZeroAddress, fmt.Errorf("split metadata error: %v", err)
+	}
+
+	// now join both references (mr, fr) to create an entry and store it
```

Comment:
I also left this copy/paste (order in comment is different). Just mentioning it here, my plan was to create small PR for these typos later.

Reply:
sure, i only made minor aesthetic changes when extracting this func, but thanks for the heads-up